rust/kernel/io_mem.rs

```diff
@@ -177,6 +177,47 @@ impl<const SIZE: usize> IoMem<SIZE> {
         crate::build_assert!(Self::offset_ok::<T>(offset), "IoMem offset overflow");
     }
 
+    /// Copy memory block from an i/o memory into a buffer.
+    ///
+    /// # Examples
+    /// ```
+    /// # use kernel::prelude::*;
+    /// use kernel::io_mem::{self, IoMem, Resource};
+    ///
+    /// fn test(res: Resource) -> Result {
+    ///     // Create an i/o memory block of at least 100 bytes.
+    ///     let mem = unsafe { IoMem::<100>::try_new(res) }?;
+    ///
+    ///     let mut buffer: [i8; 32] = [0; 32];
+    ///
+    ///     // Memcpy 16 bytes from offset 10 of i/o memory block into the buffer.
+    ///     mem.try_memcpy_fromio(&mut buffer, 10, 16)?;
+    ///
+    ///     Ok(())
+    /// }
+    /// ```
+    pub fn try_memcpy_fromio(&self, from: &mut [i8], offset: usize, count: usize) -> Result {
+        if !IoMem::<SIZE>::offset_ok::<u8>(offset) {
+            return Err(Error::EINVAL);
+        }
+
+        // Check if memory block can be copied from i/o region
+        // and also whether it will fit into data buffer.
+        if count > from.len() || count > SIZE - offset {
+            return Err(Error::EINVAL);
+        }
+
+        let ptr = self.ptr.wrapping_add(offset);
+
+        // SAFETY:
+        //   - The type invariants guarantee that `ptr` is a valid pointer.
+        //   - The `offset_ok` check returns an error if `offset` would make the write
+        //     go out of bounds (including the type size).
+        //   - the sizes of from/to bufferso are asserted whether they match `count` above.
+        unsafe { bindings::memcpy_fromio(from.as_mut_ptr() as _, ptr as _, count as _) };
```

Comment:
Maybe use `as *mut _` and `as *const _` instead of `as _`?

Reply:
Agree, thanks!